### PR TITLE
[Refactor] simplify multimodal data processing

### PR DESCRIPTION
**Pull request by JustinTong0323**
This pull request aims to refactor and simplify the multimodal data processing within the system. The core change involves centralizing model-specific data into a single dictionary within the MultimodalDataItem class, reducing boilerplate and improving extensibility. This refactoring impacts how various multimodal models and their processors handle and access their specific input features.

Highlights

- Centralized Multimodal Data Storage: The MultimodalDataItem class has been refactored to consolidate model-specific attributes (e.g., image_grid_thw, audio_feature_lens) into a new model_specific_data dictionary, simplifying the data structure and making it more generic.
- Streamlined Multimodal Processors: Multimodal processors (e.g., CLIP, Deepseek-VL, Mllama) have been updated to utilize the new model_specific_data structure, reducing redundant code and improving consistency in how data is prepared for models.
- Model Adaptations: Numerous model implementations (e.g., Deepseek-VL2, Gemma3N, Kimi-VL, MiniCPM, Phi4MM, Qwen2.5-VL, Qwen2-Audio, Qwen2-VL) have been adjusted to retrieve their specific input features from the model_specific_data dictionary.
- Processor Utility Enhancements: The BaseMultimodalProcessor now includes improved logic for collecting and combining multimodal data, leveraging the new centralized data structure.
- File Renaming/Restructuring: The qwen_audio.py processor file has been moved and renamed, indicating a potential restructuring of the multimodal processor directory.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/mllama.py [hunk 1]
```diff
@@ -836,19 +836,10 @@ def __init__(
             prefix="multi_modal_projector",
         )
         self.logits_processor = LogitsProcessor(config.text_config)
+        self.padding_pattern = MultiModalityDataPaddingPatternMultimodalTokens()
 
     def pad_input_ids(self, input_ids: List[int], mm_inputs: MultimodalInputs):
```
Comment (mickqian) [inline python/sglang/srt/models/mllama.py]: we can move `pad_input_ids` into processor in another PR

**Comment by mickqian** [comment]
this has quite some duplication with #7924. Can we prioritize one?

**Comment by JustinTong0323** [comment]
> this has quite some duplication with #7924. Can we prioritize one?

Yes we could first prioritize that PR

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -211,37 +211,11 @@ class MultimodalDataItem:
     # the raw features returned by processor, e.g. pixel_values or audio_features
     feature: Union[torch.Tensor, np.ndarray] = None
 
-    image_sizes: Tuple[int, int] = None
+    # the precomputed embeddings for the modality, e.g. image_emb for image, audio_emb for audio
+    precomputed_embeddings: Optional[Union[torch.Tensor, np.ndarray]] = None
```
Comment (mickqian) [inline python/sglang/srt/managers/schedule_batch.py:215]: maybe plural is not needed?

File: python/sglang/srt/multimodal/processors/janus_pro.py [hunk 3]
```diff
@@ -11,52 +11,35 @@
...
+            image_token=_processor.image_token,
+            image_token_id=_processor.image_id,
+        ).build(_processor)
 
     async def process_mm_data_async(
         self,
         image_data: List[Union[str, bytes]],
         input_text,
         request_obj,
-        max_req_input_len,
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/janus_pro.py:26]: We might still need this, which is useful for controlling input len

File: python/sglang/srt/multimodal/processors/mlama.py [hunk 4]
```diff
@@ -1,34 +1,38 @@
...
 
 
 class MllamaImageProcessor(BaseMultimodalProcessor):
     models = [MllamaForConditionalGeneration]
 
     def __init__(self, hf_config, server_args, _processor):
         super().__init__(hf_config, server_args, _processor)
+        self.IM_TOKEN = self._processor.image_token
+        self.IM_TOKEN_ID = self._processor.image_token_id
+        self.mm_tokens = MultimodalSpecialTokens(
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/mlama.py:15]: nit: I would prefer skip storing `self.IM_TOKEN_ID`  itself. a variable in `mm_tokens` is enough, and it's easier if we want to return mm_tokens in the future

File: python/sglang/srt/managers/schedule_batch.py [hunk 5]
```diff
@@ -306,8 +269,7 @@ def from_dict(obj: dict):
 
     def merge(self, other):
         self.feature += other.feature
-        self.image_sizes += other.image_sizes
-        self.image_offsets += other.image_offsets
+        self.offsets += other.offsets
         self.hash = hash((self.hash, other.hash))
         self.set_pad_value()
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply python/sglang/srt/managers/schedule_batch.py:294]: @mickqian What do you think about it? Maybe could resolve it when merge is used?

File: python/sglang/srt/multimodal/processors/janus_pro.py (hunk 3 above)
Comment (JustinTong0323) [reply python/sglang/srt/multimodal/processors/janus_pro.py:26]: Currently only deepseek vl2 use it and I remain as a optional kwarg.

File: python/sglang/srt/models/llava.py [hunk 6]
```diff
@@ -753,7 +759,9 @@ def get_image_feature(self, items: List[MultimodalDataItem]) -> torch.Tensor:
         features = []
         for item in items:
             # in each item, we assume pixel_values is always batched
-            pixel_values, image_sizes = item.feature, item.image_sizes
+            pixel_values, image_sizes = item.feature, item.model_specific_data.get(
```
Comment (mickqian) [inline python/sglang/srt/models/llava.py]: can we add a `get` method to `MultimodalDataItem`?

File: python/sglang/srt/managers/schedule_batch.py (hunk 2 above)
Comment (JustinTong0323) [reply python/sglang/srt/managers/schedule_batch.py:215]: I'm not sure, as the embedding could represent multiple images, I think remain plural here is more general

File: python/sglang/srt/models/llava.py (hunk 6 above)
Comment (JustinTong0323) [reply python/sglang/srt/models/llava.py]: DONE

File: python/sglang/srt/managers/schedule_batch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/srt/managers/schedule_batch.py:294]: Do we have the merge logic for these field before? if not, it is not urgent

File: python/sglang/srt/models/llava.py (hunk 6 above)
Comment (mickqian) [reply python/sglang/srt/models/llava.py]: nit: sorry for letting you implement `get` earlier, but do you think this way it can be simpler? :
```python
class Item:
    def __getattr__(self, name):
        try:
            return self._data[name]
        except KeyError:
            raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")

# instead of `.get("some_fields")`
Item().some_fields
```

**Review by mickqian — APPROVED** [review]

File: test/srt/test_vision_openai_server_common.py [hunk 7]
```diff
@@ -308,19 +308,35 @@ def test_video_images_chat_completion(self):
             "iPod" in video_response
             or "device" in video_response
             or "microphone" in video_response
-        ), video_response
+        ), f"""
+        ====================== video_response =====================
+        {video_response}
+        ===========================================================
+        should contain 'iPod' or 'device' or 'microphone'
```
Comment (mickqian) [inline test/srt/test_vision_openai_server_common.py:315]: better not hard code this
